src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,105 @@
+#include "engine/effects/engineeffectsdelay.h"
+
+#include "util/sample.h"
+
+namespace {
+// See enginedelay.cpp
+const int kiMaxDelay = static_cast<int>(0.508 *
+        mixxx::audio::SampleRate::kValueMax * mixxx::kEngineChannelCount);
+} // anonymous namespace
+
+EngineEffectsDelay::EngineEffectsDelay()
+        : m_delaySamples(0),
+          m_oldDelaySamples(0),
+          m_delayPos(0) {
+    m_pDelayBuffer = SampleUtil::alloc(kiMaxDelay);
+}
+
+void EngineEffectsDelay::process(const CSAMPLE* M_RESTRICT pIn,
+        CSAMPLE* M_RESTRICT pOutput,
+        const int iBufferSize) {
+    if (m_oldDelaySamples == 0 && m_delaySamples == 0) {
+        SampleUtil::copy(pOutput, pIn, iBufferSize);
+        return;
+    }
+
+    if (m_oldDelaySamples == m_delaySamples) {
+        int delaySourcePos = (m_delayPos + kiMaxDelay - m_delaySamples) % kiMaxDelay;
+
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos >= 0) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos <= static_cast<int>(kiMaxDelay)) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+
+        for (int i = 0; i < iBufferSize; ++i) {
+            // Put samples into delay buffer.
+            m_pDelayBuffer[m_delayPos] = pIn[i];
+            m_delayPos = (m_delayPos + 1) % kiMaxDelay;
+
+            // Take a delayed sample from the delay buffer
+            // and copy it to the destination buffer.
+            pOutput[i] = m_pDelayBuffer[delaySourcePos];
+            delaySourcePos = (delaySourcePos + 1) % kiMaxDelay;
+        }
```

Comment:
if you feel lucky you may fiddle around to either use factorized functions or make this loop vetcorized. 
This can be checked like described here: https://github.com/mixxxdj/mixxx/blob/0ad01b6ccc744045fb727eb4e775aa0e52bf4998/src/util/sample.cpp#L14

This requires that you chop the loop in chunks to get around the % kiMaxDelay on every sample.